### ISSUE TITLE
Load images from IPFS

### ISSUE THREAD

**Issue by geir54** [Gargron/mastodon#360]
It would be great if mastodon was using [ipfs](https://ipfs.io/) to load/store images. This way if a server goes down the content would stay up if someone else has loaded it. It would also reduce the bandwidth on the server hosting the original content.

**Comment by ineffyble** [comment]
Related to #477 but has a more limited scope (just media), which sounds more feasible, as alternate media backends already exist (i.e. S3)

**Comment by ProgVal** [comment]
Although I'm a big fan of the ideas in IPFS, I would prefer it not to be a depdency of Mastodon.
Mastodon already depends heavily on Ruby and Nodejs; adding IPFS would significantly complexify instance creation because it requires a Golang environment and a rather huge set of dependencies.

**Comment by mecab** [comment]
I'm now adding IPFS media backend support by writing custom Paperclip Storage ([mecab/paperclip-ipfs-storage](https://github.com/mecab/paperclip-ipfs-storage)) and integrated to mastodon. The storage is built on top of [hjoest/ruby-ipfs-api](https://github.com/hjoest/ruby-ipfs-api) so it is **not** depends on Golang. 😊

If you are interested, you can try it from [my fork](https://github.com/mecab/mastodon/tree/ipfs). (note you should use `ipfs` branch). It has an issue which is unstable connection with Docker so I will need to fix it before raising PR.

Here is the screen capture. You can see the image is served via IPFS gateway.
<img width="980" alt="screen shot 2017-05-10 at 1 27 32 am" src="https://cloud.githubusercontent.com/assets/1580772/25861541/08d30c8e-3520-11e7-879a-7923f36d462d.png">

Anyway I'm just posted here to show current progress. Any suggestions and questions are welcome 😆

Possible related issues: #477, #1847

**Comment by Gargron** [comment]
I'm worried about not being able to delete files. What does everyone else think?

**Comment by nightpool** [comment]
I would really like to take this and work on it at some point. I think that running an IPFS server alongside a mastodon instance as an optional enhancement (instead of using S3) makes a ton of sense—decentralized file storage for decentralized social media. 

@Gargron I think that not deleting things is probably a red herring as far as our application goes. If the status gets deleted then there will be no references to the media file, and there will be no way to find it except by knowing the hash. Furthermore, it can (and should) also get unpinned by any of the nodes that are pinning it, causing them to garbage collect it, meaning that the only ways it would still be accessible is if any non-mastodon IPFS nodes had mirrored it. (same as status deletion, basically)

we should *not* use the IPFS gateway though—instead, mastodon instances should have their own local gateways and we should document setting that up. There's some bikeshedding to be had here about how to make these gateways non-public (some nginx magic?)

**Comment by nightpool** [comment]
re: deletion, I'll try asking the IPFS team if they'd consider best-effort deletion as an improvement to the protocol.

**Comment by mecab** [comment]
@Gargron Yes, I understand that not being able to delete files could be a problem, but I agree to @nightpool that the files are slightly going be deleted as the hash got forgotten.

One concern is that some bad guy will spread the media's URL (i.e., with the hash) in other places after the toot has been deleted. However in this case he/she will try to paste the original content (instead of the URL) if we don't use IPFS.

**Comment by mecab** [comment]
Avoiding use of public gateway but setting up special gateway that hide the hash but it checks the availability of the original toot then proxy the contents from IPFS network could solve the problem.

**Comment by davidak** [comment]
You see often screenshots of tweets (in the media) and i download pictures that i like. So you can't really delete something from the internet.

The idea of IPFS is to archive the internet when people care about the content (e.g. pin it).

**Comment by mecab** [comment]
@davidak Yes, it is what I wanted to say. I'm for that no way to delete the media could be not a big problem.

**Comment by ProgVal** [comment]
@mecab Does your implementation support unpinning files that should be deleted?

**Comment by ProgVal** [comment]
@mecab Other question (sorry for the spam): is it possible to migrate an existing Mastodon instance to the IPFS storage, without breaking existing URLs?

**Comment by wxcafe** [comment]
Is this branch still compatible with the current state of mastodon? If not, would you be willing to port it @mecab?

**Comment by mecab** [comment]
@ProgVal Oops, I found your mention now, sorry.
> Does your implementation support unpinning files that should be deleted?
Currently not, but I think it is possible to implement.

> is it possible to migrate an existing Mastodon instance to the IPFS storage, without breaking existing URLs?
Could be. But it needs some script to done it.

**Comment by mecab** [comment]
@wxcafe > Is this branch still compatible with the current state of mastodon?

I'm sorry but I'm not sure about that since I cannot have enough time to develop recently. If IPFS support is urgent, it is no problems for me that you implement using my pieces of code, or start from scratch.

I think it needs just few modification to adapt to current state even if had conflicts. But please note I have still not resolved this (https://github.com/mecab/paperclip-ipfs-storage/issues/1) issue.

**Comment by wxcafe** [comment]
it's not urgent (last comment before mine was over 3 months ago ^^), just checking. Thanks for the work you've already done on this 👍

**Comment by mecab** [comment]
I see, thanks!

**Comment by Lionirdeadman** [comment]
IPFS should not be used in my opinion. 

I want my stuff to be deleted if I want it to be deleted (even if rogue instances or people can copy it).

Never being able to delete pictures should not be the default behavior. (As in, it shouldn't assume that I want it to be archived forever and ever using IPFS)

This could be an option in the settings if possible.

**Comment by ProgVal** [comment]
@Lionirdeadman IPFS does not automatically replicate your content. Unless someone manually pins the content, the content will only be cached temporarily by other nodes that access it.

**Comment by Lionirdeadman** [comment]
But that still means it's **forever** on the instance my account is on. Does it not?

(I may have misunderstood the talk above) 

@ProgVal

**Comment by ProgVal** [comment]
@Lionirdeadman It's forever on that instance, unless the instance unpins it. (search for "unpin" in previous messages in the thread)

**Comment by Lionirdeadman** [comment]
If I understood correctly, unpinning only forgets the *location* of the data and not the *data*. 

So my data still exists there and I don't want that.

(Please correct me if I'm wrong)

@ProgVal

**Comment by ProgVal** [comment]
unpinning causes your instance to eventually forget the data.

**Comment by Lionirdeadman** [comment]
Eventually? How much time would it generally take for it to be completely gone/overwritten/deleted after being unpinned?

I feel that if it depends on instance activity, it's a bad idea because you can't guarantee to the user that the data will be deleted in any kind of timeframe and this could be legal trouble too.

**Comment by petersjt014** [comment]
It [looks like](https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#datastore) scheduled garbage collection is disabled by default, but collection after reaching the watermark is not.

A bit on your earlier comment:

IPFS is content-addressed. When you ask the network to find a file, you are giving it the content (or at least a 'fingerprint' of it) and obtaining locations to download from. This is roughly the reverse of more traditional/common location-addressed systems like website urls.

If you add something to your repo, it's 'private' as long as only you know the hash. As soon as one other person knows, keeping it private relies on trusting them (and this doesn't apply if it's leaked in public, of course).

If others know the hash but the file is unique (meaning **only** you own it), then they can't get it unless you bring your client online while the the file is added to your repo--but if the file is small enough that someone could 'guess' it via brute force, then it's also no longer private. That probably isn't likely unless they already have a significant fragment of it and know how you chunked it (if you didn't use the default)--if not, then even 1 KiB represents 2^8192 - 1 possible files.

Nearly every [possible](https://en.wikipedia.org/wiki/The_Library_of_Babel) file has a unique hash (or at least the chances of collisions are extremely low). For an example, I hashed "lol" (with a linefeed at the end because echo does that on unix) while offline on one machine and then (on a different one) queried the DHT for the resulting hash with `ipfs dht findprovs QmQsZSD...` and found that it already existed on nine other hosts. You can do this for any possible value, but well-known and/or short values will be much more likely to pop up. 

There's more I could say here, but this comment is already pretty long. Hope this helps and that it isn't too pedantic.

**Comment by ProgVal** [comment]
> Eventually? How much time would it generally take for it to be completely gone/overwritten/deleted after being unpinned?
> 
> I feel that if it depends on instance activity, it's a bad idea because you can't guarantee to the user that the data will be deleted in any kind of timeframe and this could be legal trouble too.

The instance can trigger a flush of its cache (or only remove your content from the cache)

> If you add something to your repo, it's 'private' as long as only you know the hash. As soon as one other person knows, keeping it private relies on trusting them (and this doesn't apply if it's leaked in public, of course).

No. As for any system with a DHT, IPFS "gossips", which allows other people in the network to get the hash of the content.

**Comment by ProgVal** [comment]
> Eventually? How much time would it generally take for it to be completely gone/overwritten/deleted after being unpinned?
> 
> I feel that if it depends on instance activity, it's a bad idea because you can't guarantee to the user that the data will be deleted in any kind of timeframe and this could be legal trouble too.

The instance can trigger a flush of its cache (or only remove your content from the cache)

> If you add something to your repo, it's 'private' as long as only you know the hash. As soon as one other person knows, keeping it private relies on trusting them (and this doesn't apply if it's leaked in public, of course).

No. As for any system with a DHT, IPFS "gossips", which allows other people in the network to get the hash of the content.

**Comment by Lionirdeadman** [comment]
Why have this in the first place if other hosts can simply cache the image and content? Why use IPFS if it can leak data (which may or may not be personal in the case of a direct toot)?

I feel there is little benefit and that users should be in control of whether or not they want their data handled this way.

**Comment by ProgVal** [comment]
> Why have this in the first place if other hosts can simply cache the image and content?

Why have what? pinning?
Pinning makes sure that even if the original uploader disappears, someone else will still have it, even if the content hasn't been accessed in a while.

> Why use IPFS if it can leak data (which may or may not be personal in the case of a direct toot)?

IPFS can deduplicate content across instances and prevents an instance from being a single point of failure.

> I feel there is little benefit and that users should be in control of whether or not they want their data handled this way.

Yes. One possibility would be to only push content to IPFS if the toot is public (or unlisted?)

**Comment by zumbatan** [comment]
@Gargron  and @others.  who are taking a liking into the ipfs idea.
please consider issue #6898   .

as this is highly new and innovative space which we have here which changes dynamically (blockchain/ipfs) please I suggest you and everyone else who sees any kind of potential in my idea/vision , to contact directly the SHift team, i am not Shift. I just had this spotaneous in my view highly valuable idea to make mastodon future proof and even more decentralized and secure by leveraging anyhow Shift`s tech.

Shifts founder is called RALF and another member is called Heisenberg. both very talkable on their ryverchat:

https://shiftnrg.ryver.com/application/login

just get in touch with them for any questions you might have, you are all great Devs and understand each other effectively:)

would be nice if you updated this post of mine, whether anyone followed up and did any kind of proof of concept with Shift or ruled out the meaningfulness of it…

thx & good luck @Devs"!

**Comment by petersjt014** [comment]
Ah, right. That makes sense. I was wondering why there's almost always a trickle of traffic--that'd be the gossip then.

For distributed content replication, something key-based would probably be best. There's at least [one](http://www.bittorrent.org/beps/bep_0046.html) standard for this, and Zeronet is a working example (I think they implement the BEP or at least something like it).

**Comment by Gargron** [comment]
It seems that in many ways the current implementation acts very similar to what IPFS would offer. IPFS pinning means the original instance is still on the hook for the required storage. One benefit is file-based deduplication, but this is possible without IPFS too. Both with and without IPFS it's obviously a not perfect solution because the same image can end up with a different hash based on compression levels, dimensions, or format. There is an outstanding issue for deduplicating #2317 (which is unfortunately most difficult due to the necessary data migrations on quite large tables required). IPFS and an IPFS gateway would also be an additional deployment dependency. Closing this.

**Comment by drequivalent** [comment]
If I understand IPFS correctly, another benefit of IPFS is data locality is no longer locked to one storage provider.

You can have different storage providers for one instance, and migrate freely between them, pretty much without hassle.

**Comment by Martinligabue** [comment]
Id' also want to point out that using ipfs would lower server load, since if someone is using ipfs they'd download from other peers instead of the central server

**Comment by tfardet** [comment]
Given the current growth of Mastodon instances and activity on the network (especially with people coming with habits from Twitter, this issue and #2317 might be a good way to improve server resilience and reduce the platform's footprint

**Comment by ineffyble** [comment]
From my observations, images are not a huge _performance_ issue when it comes to scaling, because there's many great options to offload that capacity (S3 and compatible hosts, CDNs, etc). However the storage and bandwidth scaling does seem to potentially create _financial_ challenges (most CDNs don't have a huge free tier, and S3 storage fees can rack up quickly on a larger instance).
